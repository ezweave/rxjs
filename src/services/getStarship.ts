import axios, { Axios } from 'axios';
import { flow } from 'lodash';
import { map as mapFP } from 'lodash/fp';

import { People } from '@rx-example/types';

import { getStarshipURLFromPeople } from './getStarshipURLFromPeople';

interface GetStarship {
  (person: People): Promise<any>[];
}

export const getStarship = (httpGet: Axios['get'] = axios.get): GetStarship =>
  flow(
    getStarshipURLFromPeople,
    mapFP(async (url: string) => {
      const starship = await httpGet<any>(url);
      return starship.data;
    }),
  );
