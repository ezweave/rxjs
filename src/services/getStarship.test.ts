import {
  ImperialShuttleJSON,
  LukeSkywalkerPeople,
  XWingJSON,
} from './__fixtures__';
import { getStarship } from './getStarship';

describe(getStarship, () => {
  it('gets starships for each url', async () => {
    const mockGet = jest
      .fn()
      .mockResolvedValueOnce({ data: XWingJSON })
      .mockResolvedValueOnce({ data: ImperialShuttleJSON });
    const starships = await Promise.all(
      getStarship(mockGet)(LukeSkywalkerPeople),
    );
    expect(starships).toMatchSnapshot();
  });
});
