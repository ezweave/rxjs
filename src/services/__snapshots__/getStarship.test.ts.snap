// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getStarship gets starships for each url 1`] = `
Array [
  Object {
    "MGLT": "100",
    "cargo_capacity": "110",
    "consumables": "1 week",
    "cost_in_credits": "149999",
    "created": "2014-12-12T11:19:05.340000Z",
    "crew": "1",
    "edited": "2014-12-20T21:23:49.886000Z",
    "films": Array [
      "https://swapi.dev/api/films/1/",
      "https://swapi.dev/api/films/2/",
      "https://swapi.dev/api/films/3/",
    ],
    "hyperdrive_rating": "1.0",
    "length": "12.5",
    "manufacturer": "Incom Corporation",
    "max_atmosphering_speed": "1050",
    "model": "T-65 X-wing",
    "name": "X-wing",
    "passengers": "0",
    "pilots": Array [
      "https://swapi.dev/api/people/1/",
      "https://swapi.dev/api/people/9/",
      "https://swapi.dev/api/people/18/",
      "https://swapi.dev/api/people/19/",
    ],
    "starship_class": "Starfighter",
    "url": "https://swapi.dev/api/starships/12/",
  },
  Object {
    "MGLT": "50",
    "cargo_capacity": "80000",
    "consumables": "2 months",
    "cost_in_credits": "240000",
    "created": "2014-12-15T13:04:47.235000Z",
    "crew": "6",
    "edited": "2014-12-20T21:23:49.900000Z",
    "films": Array [
      "https://swapi.dev/api/films/2/",
      "https://swapi.dev/api/films/3/",
    ],
    "hyperdrive_rating": "1.0",
    "length": "20",
    "manufacturer": "Sienar Fleet Systems",
    "max_atmosphering_speed": "850",
    "model": "Lambda-class T-4a shuttle",
    "name": "Imperial shuttle",
    "passengers": "20",
    "pilots": Array [
      "https://swapi.dev/api/people/1/",
      "https://swapi.dev/api/people/13/",
      "https://swapi.dev/api/people/14/",
    ],
    "starship_class": "Armed government transport",
    "url": "https://swapi.dev/api/starships/22/",
  },
]
`;
